File: packages/ios/TimeRangeHelper.swift
import Foundation

/// Helper to manage time periods
class TimeRangeHelper {
  
  /// Period type
  enum TimeRange: String {
    case today = "today"
    case yesterday = "yesterday"
    case thisWeek = "this_week"
    case lastWeek = "last_week"
    case thisMonth = "this_month"
    case lastMonth = "last_month"
    case thisYear = "this_year"
    case lastYear = "last_year"
    case last7Days = "last_7_days"
    case last30Days = "last_30_days"
    case last90Days = "last_90_days"
    case custom = "custom"
  }
  
  /// Returns start and end dates for a given period
  static func getDates(for timeRange: TimeRange, customStart: Date? = nil, customEnd: Date? = nil) -> (start: Date, end: Date) {
    let calendar = Calendar.current
    let now = Date()
    // Safe fallback when a Calendar computation returns nil (extremely rare).
    let today = (start: calendar.startOfDay(for: now), end: now)

    switch timeRange {
    case .today:
      return today

    case .yesterday:
      guard let yesterday = calendar.date(byAdding: .day, value: -1, to: now) else { return today }
      let start = calendar.startOfDay(for: yesterday)
      guard let end = calendar.date(byAdding: .day, value: 1, to: start) else { return today }
      return (start, end)

    case .thisWeek:
      guard let start = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: now)) else { return today }
      return (start, now)

    case .lastWeek:
      guard let thisWeekStart = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: now)),
            let lastWeekStart = calendar.date(byAdding: .weekOfYear, value: -1, to: thisWeekStart) else { return today }
      return (lastWeekStart, thisWeekStart)

    case .thisMonth:
      guard let start = calendar.date(from: calendar.dateComponents([.year, .month], from: now)) else { return today }
      return (start, now)

    case .lastMonth:
      guard let thisMonthStart = calendar.date(from: calendar.dateComponents([.year, .month], from: now)),
            let lastMonthStart = calendar.date(byAdding: .month, value: -1, to: thisMonthStart) else { return today }
      return (lastMonthStart, thisMonthStart)

    case .thisYear:
      guard let start = calendar.date(from: calendar.dateComponents([.year], from: now)) else { return today }
      return (start, now)

    case .lastYear:
      guard let thisYearStart = calendar.date(from: calendar.dateComponents([.year], from: now)),
            let lastYearStart = calendar.date(byAdding: .year, value: -1, to: thisYearStart) else { return today }
      return (lastYearStart, thisYearStart)

    case .last7Days:
      guard let start = calendar.date(byAdding: .day, value: -7, to: now) else { return today }
      return (start, now)

    case .last30Days:
      guard let start = calendar.date(byAdding: .day, value: -30, to: now) else { return today }
      return (start, now)

    case .last90Days:
      guard let start = calendar.date(byAdding: .day, value: -90, to: now) else { return today }
      return (start, now)

    case .custom:
      guard let customStart = customStart, let customEnd = customEnd else {
        return today
      }
      return (customStart, customEnd)
    }
  }
  
  /// Create a readable description of the period
  static func description(for timeRange: TimeRange, start: Date? = nil, end: Date? = nil) -> String {
    let formatter = DateFormatter()
    formatter.dateStyle = .medium
    formatter.timeStyle = .none
    
    switch timeRange {
    case .today:
      return "Today"
    case .yesterday:
      return "Yesterday"
    case .thisWeek:
      return "This week"
    case .lastWeek:
      return "Last week"
    case .thisMonth:
      return "This month"
    case .lastMonth:
      return "Last month"
    case .thisYear:
      return "This year"
    case .lastYear:
      return "Last year"
    case .last7Days:
      return "Last 7 days"
    case .last30Days:
      return "Last 30 days"
    case .last90Days:
      return "Last 90 days"
    case .custom:
      if let start = start, let end = end {
        return "\(formatter.string(from: start)) - \(formatter.string(from: end))"
      }
      return "Custom period"
    }
  }
}

File: packages/android/src/main/kotlin/io/github/n0ku/nitrohealthkit/util/TimeRangeHelper.kt
package io.github.n0ku.nitrohealthkit.util

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.YearMonth
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters

/**
 * Mirrors `ios/TimeRangeHelper.swift`: given a logical range identifier
 * (`today`, `this_week`, `last_30_days`, `custom`, ...), produces a concrete
 * [startInstant, endInstant) pair anchored on the device's default timezone.
 *
 * The values are kept identical to the [com.margelo.nitro.healthkit.HybridHealthKitSpec]
 * contract on the TS side (see [src/specs/HealthKit.nitro.ts]: `enum TimeRange`).
 */
object TimeRangeHelper {

    data class Range(val start: Instant, val end: Instant)

    fun resolve(
        rangeKey: String,
        customStart: Instant?,
        customEnd: Instant?,
        zone: ZoneId = ZoneId.systemDefault(),
        now: LocalDateTime = LocalDateTime.now(zone),
    ): Range {
        val today: LocalDate = now.toLocalDate()
        return when (rangeKey.lowercase()) {
            "today" -> dayRange(today, zone)
            "yesterday" -> dayRange(today.minusDays(1), zone)
            "this_week" -> weekRange(today, zone)
            "last_week" -> weekRange(today.minusWeeks(1), zone)
            "this_month" -> monthRange(YearMonth.from(today), zone)
            "last_month" -> monthRange(YearMonth.from(today).minusMonths(1), zone)
            "this_year" -> yearRange(today.year, zone)
            "last_year" -> yearRange(today.year - 1, zone)
            "last_7_days" -> rollingDays(today, 7, zone)
            "last_30_days" -> rollingDays(today, 30, zone)
            "last_90_days" -> rollingDays(today, 90, zone)
            "custom" -> Range(
                start = customStart ?: dayRange(today, zone).start,
                end = customEnd ?: dayRange(today, zone).end,
            )
            else -> throw IllegalArgumentException("Unknown time range: $rangeKey")
        }
    }

    private fun dayRange(day: LocalDate, zone: ZoneId): Range = Range(
        start = day.atStartOfDay(zone).toInstant(),
        end = day.atTime(LocalTime.MAX).atZone(zone).toInstant(),
    )

    private fun weekRange(reference: LocalDate, zone: ZoneId): Range {
        val monday = reference.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        val sunday = reference.with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY))
        return Range(
            start = monday.atStartOfDay(zone).toInstant(),
            end = sunday.atTime(LocalTime.MAX).atZone(zone).toInstant(),
        )
    }

    private fun monthRange(month: YearMonth, zone: ZoneId): Range = Range(
        start = month.atDay(1).atStartOfDay(zone).toInstant(),
        end = month.atEndOfMonth().atTime(LocalTime.MAX).atZone(zone).toInstant(),
    )

    private fun yearRange(year: Int, zone: ZoneId): Range = Range(
        start = LocalDate.of(year, 1, 1).atStartOfDay(zone).toInstant(),
        end = LocalDate.of(year, 12, 31).atTime(LocalTime.MAX).atZone(zone).toInstant(),
    )

    private fun rollingDays(today: LocalDate, days: Long, zone: ZoneId): Range {
        val start = today.minusDays(days - 1).atStartOfDay(zone).toInstant()
        val end = today.atTime(LocalTime.MAX).atZone(zone).toInstant()
        return Range(start = start, end = end)
    }
}
